packages/front-end/components/DataViz/DataVizFilter.tsx

```diff
@@ -0,0 +1,472 @@
+  rows: Record<string, unknown>[],
+  columnName: string,
+): string[] {
+  const uniqueValues = new Set<string>();
+
+  rows.forEach((row) => {
+    const value = row[columnName];
+    if (value != null) {
+      uniqueValues.add(String(value));
+    }
+  });
+
+  return Array.from(uniqueValues);
+}
+
+const filterOptions = [
+  // Date filters
+  { value: "today", label: "Today", supportedTypes: ["date"] },
+  { value: "last7Days", label: "Last 7 Days", supportedTypes: ["date"] },
+  { value: "last30Days", label: "Last 30 Days", supportedTypes: ["date"] },
+  { value: "dateRange", label: "Custom Date Range", supportedTypes: ["date"] },
+
+  // Number filters
+  { value: "numberRange", label: "Custom Range", supportedTypes: ["number"] },
+  { value: "greaterThan", label: "Greater Than", supportedTypes: ["number"] },
+  {
+    value: "greaterThanOrEqualTo",
+    label: "Greater Than or Equal To",
+    supportedTypes: ["number"],
+  },
+  { value: "lessThan", label: "Less Than", supportedTypes: ["number"] },
+  {
+    value: "lessThanOrEqualTo",
+    label: "Less Than or Equal To",
+    supportedTypes: ["number"],
+  },
+  { value: "equalTo", label: "Equals", supportedTypes: ["number"] },
+
+  // String filters
+  { value: "includes", label: "Select Values", supportedTypes: ["string"] },
+  { value: "contains", label: "Text Search", supportedTypes: ["string"] },
+];
+
+export default function DataVizFilter({
+  filterIndex,
+  dataVizConfig,
+  onDataVizConfigChange,
+  rows,
+  columnFilterOptions,
+}: Props) {
+  const filters = dataVizConfig.filter || [];
+
+  const updateFilter = (newFilter: FilterConfiguration) => {
+    const newFilters = [...filters];
+    newFilters[filterIndex] = newFilter;
+    onDataVizConfigChange({
+      ...dataVizConfig,
+      filter: newFilters,
+    });
+  };
+
+  const updateFilterConfig = (
+    configUpdates: Record<string, string | number | string[] | undefined>,
+  ) => {
+    const currentFilter = filters[filterIndex];
+    const currentConfig = currentFilter?.config || {};
+
+    // Handle undefined values by deleting the key
+    const newConfig = { ...currentConfig };
+    Object.entries(configUpdates).forEach(([key, value]) => {
+      if (value === undefined) {
+        delete newConfig[key];
+      } else {
+        newConfig[key] = value;
+      }
+    });
+
+    // Create updated filter with new config, maintaining discriminated union structure
+    const updatedFilter: FilterConfiguration = {
+      ...currentFilter,
+      config: newConfig,
+    } as FilterConfiguration;
+
+    updateFilter(updatedFilter);
+  };
+
+  const removeFilter = () => {
+    const newFilters = [...filters];
+    newFilters.splice(filterIndex, 1);
+    onDataVizConfigChange({
+      ...dataVizConfig,
+      filter: newFilters,
+    });
+  };
+
+  const createDefaultFilterForType = (
+    column: string,
+    type: "string" | "number" | "date",
+  ): FilterConfiguration => {
+    if (type === "date") {
```

Comment:
suggestion: Can we do a exhaustive switch here?

The goal is to make TypeScript complain if we add a new filter type but don't update the function. Currently if we add a new one it would be treated as string.

Reply:
Took a pass at this in this commit.

It's definitely better and the `type` options are now based on `FilterConfiguration["type"]` so only known types can be passed in.

But I'm not sure of a way to make it more exhaustive outside of writing some custom logic to make sure each type is accounted for. If you do know of a way, I'd love to implement it.

Reply:
TIL!